### ISSUE TITLE
Should we merge `export` command into get?

### ISSUE THREAD

**Issue by b5** [qri-io/qri#1176]
From discord chat:

> wondering  whether qri export has a future in general. The differences from qri checkout seem nuanced. (I think just that checkout also links the resulting output dir? Is there anything else?) It seems like you could accomplish the same with several calls to qri get. Not sure whether it’s an important enough use-case to have a shortcut that might be confusing to many users.
- @Mr0grog 

> It keeps being a problem that export is misunderstood which is a sign that it's not pulling its own weight
- @dustmop 

RFC0014 defined a need for Qri Export:
https://github.com/qri-io/rfcs/blob/master/text/0014-export.md

Then we shipped support for File System integration in RFC0025, which solved a lot of use cases export used to cover.

What If we merged all of export's functionality into `get`?

Qri Desktop relies on the `/export` endpoint to get versions, we'd need to migrate that to use get.

**Comment by dustmop** [comment]
In my opinion, the RFC that is most relevant to this question is this one:
https://github.com/qri-io/rfcs/blob/master/text/0021-export_behavior.md

As this was written before we made FSI (file system integration), we can see that much of the "guide-level discussion" is no longer applicable. Most of these questions of allowing qri datasets to exist outside of our repository in a different representation is now easily doable by using `qri checkout`. The remaining capabilities could all be integrated into `qri get`, by allowing conversion into other file format. Even the high-fidelity export, which we never fully implemented, could be folded in as a new format that we define.

Export isn't doing enough work on its own to continue to exist, considering all of the confusion it causes.

> Qri Desktop relies on the `/export` endpoint to get versions, we'd need to migrate that to use get.

Is this true? I thought that the `/export` endpoint always writes to disk, and doesn't return any body. I thought it was only using `/export` for right-clicking and choosing "export".

**Comment by Mr0grog** [comment]
Reading these RFCs and playing with the command, I now realize my mental model for what this command does is *way* off from reality. I think my understanding was fairly close to RFC 14, but actual functionality is closer to RFC 21, and is very different.

For instance, I thought these two commands were pretty close to the same:

```sh
# A basic checkout:
$ qri checkout me/awesome_data

# vs. export + link:
$ qri export -o awesome_data me/awesome_data
$ cd awesome_data
$ qri fsi link me/awesome_data
$ cd ..
```

But they really aren’t. It might be a closer match to `qri export me/awesome_data --takeout` as described in RFC 21, which does not (yet?) exist. In any case, I think I understand why @dustmop pushed back on my thought that resembled a shortcut for several `get` calls now. 😬

(Side note: the CLI help says “To export to a specific directory, use the --output flag” but I think `--output` actually always has to be a single *file*. It might be part of what led me down the wrong conceptual path.)

Long story short, I think I may have been at least a little wrong-headed in suggesting that the functionality was redundant and added unnecessary complexity. My misunderstanding was at least partly a factor of documentation.

**Comment by Mr0grog** [comment]
That said, if this is still worth discussing, maybe it’s worth returning to the original goals from the RFCs here?

1. Support interop with other tools, e.g. Excel.
2. Getting a full-fidelity copy of a dataset disentangled from qri itself. (And in a form you don’t need qri to read.)
3. Creating a portable copy of the dataset that can be passed around or archived, then re-imported later (with a hypothetical future `qri import` command).
4. Putting the entirety of the dataset into a single file for easier analysis with other tools and scripts.

(1) Is semi-solved by `qri checkout` and that gives you an easier path to updating the dataset after editing with your “other tool.” (One BIG caveat is that checkout can’t convert formats, e.g. to give you a JSON file if the original/native form of the dataset was XLSX. Should `export` be a solution for that? Should you use `qri get body --format` after checking out instead? Is this a hole that needs filling?)

(2) Seems like a really nice feature to have, and is what I understood the goal of `export` to be (though I understood it in a much narrower form than it actually is).
- (2) also happens to align well with (3), but only if the formats are more limited. I imagine doing so would also really reduce code complexity.
- As a user, I *think* I would naturally expect/desire (2) and (3) to work together.
    - Both RFCs make big caveats out of the fact that many options in this feature are something of a foot-gun with respect to (3).
- How harmful to (2) would it be to limit the formats so it always works with (3)?
    - I think (?) `json` would probably not serve (2) well enough — you could read it without qri, but the structure would be *really* qri-specific. The directory (or zip of a directory) form gives you the body in a more standard/interoperable form.
- I can also achieve both (2) and (3) by checking out a dataset and unlinking it. Is that clunky enough and this use-case important enough to have an export command? (I think yes…?)

(3) See above.

(4) Fits with *one* of the possible forms of (2) and (3), and so maybe also aligns well with more limited export options. But I think `checkout` was also intended to make scripting or other outside tools fit with qri better, so I’m not sure if this use-case is still considered important.

This framing feels like it points to `qri export` still being useful, but that it could be stripped down a bit.

**Comment by b5** [comment]
I think your four-point breakdown is spot-on, and that export should focus on solving only 2 & 3. 

I'm a big fan of your description of how checkout solves many of the problems of integration, and agree that we should focus on expanding that functionality instead of shoehorning more stuff into export. As an example, moving `xlsx` interop work over to the checkout side is a no-brainer. I agree it's missing functionality.

I'm also digging on the idea of skinning export to _only_ serving goals 2 & 3, and think we should head in that direction. We should also adjust the language around export to clarify "you probably want to look at get & checkout instead of export".

**Comment by Mr0grog** [comment]
👍 Just don’t let me convince you to tear away any important features/use-cases you need to support. I have way less context on everything in Qri than you both of you do, and the biggest point here is really that my original concern (that kicked off this issue) stemmed from a misunderstanding of the existing functionality and goals.

**Comment by dustmop** [comment]
Coming back to this issue, wanted to add a bit more background before suggesting the next steps.

Another historical motivation that isn't really mentioned in the RFC is the distinction between internal and external representation. We used to strongly emphasize that we had an idea of a "pure" form of a dataset, something that strictly followed qri's opinionated approach to how datasets must be serialized in the repository. For example, they are immutable, they follow the model in dataset/dataset.go, they have a uniquely identifiable hash, they point to their own history, etc. The command `get` was guaranteed to operate on this "pure" form, and likewise `save` would use these invariants to create commits moving forward in time. On the other hand `export` would create an externalized representation that didn't have these assurances. This external form, living on a user's normal filesystem, could be re-imported using `save` to construct commits from the past, which was a unique consequence of how `export` tried to preserve content-awareness outside the repository.

With the `checkout` command, all of that paradigm has gone out the window, and no longer applies to help differentiate what `export` is for. We now have a new way of thinking about the filesystem, outlined in [RFC 0026](https://github.com/qri-io/rfcs/pull/48). This obsolescence is yet another reason why we should remove the `export` command, and move the relevant functionality (zipped component files, format conversion, raw block takeout) into `get`. I think all of the (1..4) use cases mentioned above can all be merged into `get`. So, what I would suggest:

----
```
qri get --format zip me/my_ds
```
Desktop needs this to support the export action, so this is high-priority to put in place. It should create a zip that contains all the components as individual files, like how checkout looks. Handles case (1).

----

```
qri get --format full-raw me/my_ds
```
To get the "takeout" version with raw blocks (flag value is just a suggestion), handles case (2) and possibly (3). We can implement this eventually.

----
```
qri get --format xlxs me/my_ds
```
This should work. I wonder if the default format should match the imported format.

----

```
qri checkout --format xlsx me/my_ds
```
Would be great to have as well.

----

I think (4) was originally talking about exporting a dataset.json with an inlined body. Might still be useful, but honestly not a huge need for this right now.

**Comment by dustmop** [comment]
I think I'd like to move forward on this, by implementing:

```
qri get --format zip me/my_ds
```

and having Desktop switch to this. Afterwards, we can remove the `export` command, and implement the other discussed features later.

**Comment by b5** [comment]
I think we should remove export for another reason: we need to nudge users to learn what `get` does early. The common understanding of "export a dataset" is "give me a CSV file". In Qri getting just the CSV file is a lossy decision, which is why it's intentionally not the default for get:

```
$ qri get body me/my_ds
```

The get command has a lot more thought behind it, and people keep using export by default, and are kinda missing the whole point of putting more than just data in datasets. `qri get body` should _feel lossy_, because it is. Removing export will drive more folks to understand that to qri a dataset is more than a structured data file. This does come with an additional documentation burden, but one I believe we're well positioned to handle.

With that said, going all-in on get brings about issues we'll need to deal with. A few points for discussion. I'm not 💯 that all this should get worked in:

### output flag
Our default output is `stdout`, but I'd like to see an `-o, --output` flag to write to a place. Many modern CLI tools acknowledge that users aren't always familiar with `$ qri get body > body.csv`, an output flag helps them wayfind.

An output flag distinguishes intent. `$ qri get me/my_ds` by should default output a YAML _dataset preview_ (currently we're still missing the body preview, but it's close), to prevent `get` from overwhelming the terminal output. Preview output is important for the common case of needing to just look at a dataset from the command line, but causes problems when get mixes with export-style intent.

When using the `-o` flag, qri should be smart, recognizing the user wants more than a preview.
I'd love an easy `get` command that mimics what `checkout` creates without needing to unzip: 

```
$ qri get me/my_ds --output .
```

should do exactly what checkout does, just without the link, writing the default FSI layout of dataset components.

getting a component with the output flag should know I'm writing to a file:

```
$ qri get meta me/some_dataset -o some_dataset_meta
```
should create `some_dataset_meta.json`


### zip flag
As for zip, I think it's a flag that wraps output in an archive, not a format. I want to be able to combine `--zip` with other things for convenience:

```
# create an archive with body.csv & qri-ref in a zip archive
$ qri get body me/my-ds --zip
```

Using the zip flag should always write a `qri-ref`, connecting back to the version it came from. Desktop will like this form more so we can build a modal that customizes "what's in the bag". eg: desktop might build a modal that let's a user "export" the metadata & body (skipping any transforms). It'll be the API-call equivalent of:

```
$ qri get meta,body  me/my_ds --zip
```

Strange caveat here, I think using `--zip` should _also_ trigger the "I want everything" intent described in the `--output` section above. `$ qri get me/my_ds --zip` should start spitting the _entire_ dataset as a zip archive  to `stdout`. That would let me [pipe to s3](https://loige.co/aws-command-line-s3-content-from-stdin-or-to-stdout/):

```
$ qri get me/my_ds --zip | aws s3 cp - s3://some-bucket/my_ds.zip
```

I'm concerned this last point makes things _very_ idiosyncratic. Would love to hear others thoughts.

### car files for "full-raw" export
Regarding high-fidelity output we're calling "full raw", our friends at protocol labs have been working on the [Content-Addressed Archive (CAR) format](https://github.com/ipld/specs/blob/master/block-layer/content-addressable-archives.md). Let's just use that when it lands in IPFS. so it'll be:

```
$ qri get me/my_ds --format car
```

**Comment by dustmop** [comment]
> I think we should remove export for another reason: we need to nudge users to learn what `get` does early

Definitely, this is the best reason. Decreasing confusion around the UI is huge for us. I've mostly been posting in this thread about why the historical justification for `export` no longer applies. 

> Our default output is stdout, but I'd like to see an -o, --output flag to write to a place.

I like this and agree.

> getting a component with the output flag should know I'm writing to a file:
> $ qri get meta me/some_dataset -o some_dataset_meta

This seems ambiguous to me. Does the user want `some_dataset_meta.json` or `some_dataset_meta/meta.json`? The previous example was using `.` which is understood to mean a directory. I feel like we should be assuming directory unless the user provides a known file extension. Thoughts?

Also, the stdout assumes format yaml, when do we switch to json instead? Anytime there's an `-o` flag?

> As for zip, I think it's a flag that wraps output in an archive, not a format. I want to be able to combine `--zip` with other things for convenience:

None of these examples would not work if you had to do `--format zip` instead. The real reason to use `--zip` instead is that you can combine it with `--format json` or `--format yaml` to change the file representations on disk. But that raises more questions, does the format apply to the components, or to the body, or both? I'm not sure the extra flag is worth the added confusion and cognitive overhead.

We should focus on what Desktop wants when it calls the API. It probably doesn't need to set both `--format json` and `--zip`, right? That makes it seem like the simpler route is better, meaning `zip` should just be a type of format. If we're opinionated about the meaning of `zip` the code will be much less complex.

> Using the zip flag should always write a qri-ref

The old file "ref.txt" existed before FSI, but since we're rewriting I think it's good to bring it into line with FSI's `.qri-ref`. Should it contain the hash and dataset ref? InitID if available? Current FSI does not contain the hash, only the ref. I think it would be good to unify the two.

> car files for "full-raw" export

This sounds cool and I like it.

**Comment by b5** [comment]
> I feel like we should be assuming directory unless the user provides a known file extension

Yes this. My version is definitely ambiguous. Sometimes I still want computers to just get it, man.

> Also, the stdout assumes format yaml, when do we switch to json instead? Anytime there's an -o flag?

This distinction is totally arbitrary, and I think that's bad, and I think we should set this in configuration, keep it consistent no matter the context, and default this to pretty-printed JSON. how about `CLI.outputFormat`? I was going to suggest defaulting to YAML, but this has horrible implications for FSI. Maybe I'm just being lazy.

>  I'm not sure the extra flag is worth the added confusion and cognitive overhead.

Yes you're totally right. I take back the `--zip` flag stuff, I'm back on team `--format zip`.

> I think it's good to bring it into line with FSI's .qri-ref.
YES PLZ.
>  Should it contain the hash and dataset ref? 
yep. 
> InitID if available?
yes... if we can figure out a syntax for ref strings that contains init IDs, which we haven't yet.

**Comment by Mr0grog** [comment]
>>> getting a component with the output flag should know I'm writing to a file:
>>> 
>>> ```
>>> $ qri get meta me/some_dataset -o some_dataset_meta
>>> ```
>>> 
>>> should create `some_dataset_meta.json`
>>
>> This seems ambiguous to me. Does the user want `some_dataset_meta.json` or `some_dataset_meta/meta.json`? The previous example was using `.` which is understood to mean a directory. I feel like we should be assuming directory unless the user provides a known file extension.
>
> Yes this. My version is definitely ambiguous. Sometimes I still want computers to just get it, man.

So just to make sure I’m following:
- The `--output/-o` flag specifies a directory
- Unless we are exporting an archive, in which case it specifies a file?


>> As for zip, I think it's a flag that wraps output in an archive, not a format. I want to be able to combine `--zip` with other things for convenience:
> 
> None of these examples would not work if you had to do `--format zip` instead. The real reason to use `--zip` instead is that you can combine it with `--format json` or `--format yaml` to change the file representations on disk. But that raises more questions, does the format apply to the components, or to the body, or both? I'm not sure the extra flag is worth the added confusion and cognitive overhead.

FWIW, I think it’s clearer/less confusion/less cognitive overhead to keep these separate, but I *might* rename `--format` to `--body-format` (sorry, I know, unenjoyably verbose, but this also separates that from however the metadata files should be formatted). Or maybe: `--format csv --archive zip`?

I also agree that needing to combine both is probably not especially common, but I can easily think of one use-case that’s not totally unusual: if I want to output a copy of a dataset to archive in some other system. If a dataset’s native format is something proprietary, like XLSX, I’d want to make sure the output body was in an [LOC-recommended archival format](https://www.loc.gov/preservation/resources/rfs/data.html):

```sh
$ qri get me/my_dataset --format csv --zip
```

> I think it's good to bring it into line with FSI's `.qri-ref`. Should it contain the hash and dataset ref? …Current FSI does not contain the hash, only the ref. I think it would be good to unify the two.

Agree!

**Comment by dustmop** [comment]
Long ago, we used to have separate `--body-format` and `--format` flags, but combined them because in nearly all cases only one has meaning; you're either getting the dataset head or the body. It's only in cases were you get both (like the export use-case) that you would need all these options, and I'm claiming it's better to totally avoid that question by reducing the number of user choices. The previous implementation of `export` was far too complex to maintain, understand, and keep correct, in part because it had way too many options. Though keeping it simple means we will miss out on providing convenience in certain cases, I don't think `qri` needs to be a swiss-army knife when it comes to one-line format conversions, as long as we provide the necessary basics that lets users combine pieces to make the tool chain that they need.

**Comment by Mr0grog** [comment]
> I'm claiming it's better to totally avoid that question by reducing the number of user choices. The previous implementation of export was far too complex to maintain, understand, and keep correct, in part because it had way too many options. Though keeping it simple means we will miss out on providing convenience in certain cases

I wrote a whole response here, but just as I was about to post, it suddenly occurred to me that the mental model you and I are thinking through might be totally different here.

**TL;DR I feel like this depends a lot on how users conceptualize what a dataset in Qri is, and I have no idea what would be most common. So probably just ignore me on this issue.**

Longer, more dawdling explanation:

I was feeling like this doesn’t reduce complexity much because `--format` will now indicate two characteristically things at the same time (i.e. data format and compression type). Under the hood, you still have to disambiguate those two concepts and treat them as separate.

But I think maybe you’re thinking of this as “transforms that have to be applied to the dataset’s files.” In other words, we’re not really implying two things at once with this option because `--format zip` doesn’t mean “default data format + zip compression;” it means “transform the raw dataset directory by zipping it.” And `--format csv` doesn’t mean “CSV data format + no compression;” it means “transform the raw dataset directory by coverting the body file from whatever it is to CSV.”

Am I getting that right? FWIW, the way I’m thinking you’re looking at it was a *huge* perspective shift for me. As a user, I generally think of Qri as holding onto some kind dataset that isn’t really files (similar to how I might think of a table in PostgreSQL, for example), and `qri get` is serializing part of it (or, now, all of it) into a file or set of files. So when I ask for `--format csv` vs. `--format xlsx`, I don’t *feel* like I’m asking Qri to do two especially different things — it’s just going from magical-internal-to-qri-data to some chosen format on disk (that also means I think of *not* setting `--format` the same as `--format <default format for this dataset>`). But I think maybe your perspective is coming more from the fact that Qri really is treating a dataset as a set of files under the hood, rather than something more special (since it uses IPFS’s UnixFS rather than a custom DAG). So `--format csv` vs. `--format xlsx` vs. no `--format` really is asking for Qri to read the underlying file and convert it if my choice differs from the file that was stored. And that operation isn’t so different from `--format zip`, where I’m asking to read out the underlying files and then zip them.

I think a lot of what’s simple or makes sense for your users hinges on how they perceive what a dataset is in Qri, and I don’t really know whether most would be more prone to thinking of it as a directory of files (where `--format csv` and `--format zip` are both asking for changes to the files of the dataset) or as something else (where `--format csv` and `--format zip` feels like two entirely different categories of things).

Anyway, I guess I have no strong opinion here now, other than: it depends a lot on how your users conceptualize what a dataset *is*. I may just be way out of sync with reality here since the docs have changed dramatically since I last really read them, and my attention and involvement in looking at Qri is not especially deep. Lots of things have changed since I first really looked at it that I don’t know about or have followed well at all.

**Comment by dustmop** [comment]
Thanks to https://github.com/qri-io/qri/pull/1453, `qri export` is no more. Use `qri get --format zip -o dataset.zip` instead.
